File: src/glean_client.py
"""
Thin HTTP clients for the Glean APIs, split by trust domain.

IndexingClient — uses the Indexing token; writes to the knowledge graph.
QueryClient    — uses the Client token; reads via Search and Chat.

This split mirrors Glean's own token model (separate Indexing vs Client tokens)
and makes the auth boundary obvious in code. See ADR-004 in DESIGN_NOTE.md.

Responsibilities:
- Attach auth headers per client.
- Retry on transient failures (429, 5xx, network) with exponential backoff.
- Emit structured logs with per-call latency and request IDs.
- Surface non-transient errors clearly.
"""
from __future__ import annotations

import json
import logging
import time
import uuid
from dataclasses import dataclass
from typing import Any, Optional

import httpx
from tenacity import (
    retry,
    retry_if_exception_type,
    stop_after_attempt,
    wait_exponential,
)

from .config import Config

log = logging.getLogger(__name__)


class GleanAPIError(Exception):
    """Non-transient API error we should surface to the caller."""

    def __init__(self, status: int, body: str, request_id: str):
        super().__init__(f"Glean API error {status} (request_id={request_id}): {body[:500]}")
        self.status = status
        self.body = body
        self.request_id = request_id


class GleanTransientError(Exception):
    """Transient error — retry candidate."""


@dataclass
class CallResult:
    status: int
    json: dict[str, Any]
    latency_ms: int
    request_id: str


class _BaseClient:
    """Shared transport layer; subclasses provide their own auth token."""

    def __init__(self, cfg: Config, token: str, label: str):
        self.cfg = cfg
        self.label = label
        self._http = httpx.Client(
            timeout=cfg.http_timeout_s,
            headers={
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json",
                "User-Agent": "glean-rag-prototype/0.1",
            },
        )

    def close(self) -> None:
        self._http.close()

    @retry(
        retry=retry_if_exception_type(GleanTransientError),
        stop=stop_after_attempt(3),
        wait=wait_exponential(multiplier=0.5, max=4),
        reraise=True,
    )
    def _post(self, url: str, payload: dict, *, op: str) -> CallResult:
        request_id = str(uuid.uuid4())
        started = time.monotonic()
        try:
            resp = self._http.post(url, json=payload)
        except httpx.RequestError as e:
            log.warning(
                "glean_api_network_error",
                extra={"client": self.label, "op": op, "url": url,
                       "request_id": request_id, "err": str(e)},
            )
            raise GleanTransientError(str(e)) from e

        latency_ms = int((time.monotonic() - started) * 1000)
        log.info(
            "glean_api_call",
            extra={
                "client": self.label,
                "op": op,
                "status": resp.status_code,
                "latency_ms": latency_ms,
                "request_id": request_id,
            },
        )

        if resp.status_code == 429 or 500 <= resp.status_code < 600:
            raise GleanTransientError(
                f"{resp.status_code} on {op} (request_id={request_id})"
            )
        if resp.status_code >= 400:
            raise GleanAPIError(resp.status_code, resp.text, request_id)

        try:
            body = resp.json()
        except json.JSONDecodeError as e:
            raise GleanAPIError(resp.status_code, resp.text, request_id) from e

        return CallResult(
            status=resp.status_code, json=body,
            latency_ms=latency_ms, request_id=request_id,
        )


class IndexingClient(_BaseClient):
    """Writes documents to the sandbox datasource via the Indexing API."""

    def __init__(self, cfg: Config):
        super().__init__(cfg, token=cfg.glean_indexing_token, label="indexing")

    def index_documents(self, documents: list[dict]) -> CallResult:
        """
        Upsert documents to the configured datasource via /indexdocuments.
        Additive — documents missing from this call are NOT removed. Use
        bulk_index_documents() for full-sync semantics.
        Note: Indexing is asynchronous — documents may not be immediately
        searchable after this call returns 200.
        """
        url = f"{self.cfg.indexing_base_url}/indexdocuments"
        payload = {
            "datasource": self.cfg.glean_datasource,
            "documents": documents,
            # uploadId makes re-runs idempotent on Glean's side too.
            "uploadId": f"prototype-{int(time.time())}",
        }
        return self._post(url, payload, op="index_documents")

    def bulk_index_documents(
        self,
        documents: list[dict],
        *,
        page_size: int = 50,
        upload_id: Optional[str] = None,
        force_restart_upload: bool = False,
        disable_stale_document_deletion_check: bool = False,
    ) -> list[CallResult]:
        """
        Full-sync upload via /bulkindexdocuments. Chunks documents into pages
        sharing one uploadId, marking isFirstPage / isLastPage. When the last
        page lands, Glean drops any documents in the datasource that weren't
        part of this upload — the canonical "replace the source" flow.
        """
        if not documents:
            raise ValueError("bulk_index_documents: documents must be non-empty")

        url = f"{self.cfg.indexing_base_url}/bulkindexdocuments"
        upload_id = upload_id or f"bulk-{int(time.time())}"
        pages = [documents[i : i + page_size] for i in range(0, len(documents), page_size)]

        results: list[CallResult] = []
        for i, page in enumerate(pages):
            is_first = i == 0
            is_last = i == len(pages) - 1
            payload: dict[str, Any] = {
                "datasource": self.cfg.glean_datasource,
                "documents": page,
                "uploadId": upload_id,
                "isFirstPage": is_first,
                "isLastPage": is_last,
            }
            if is_first and force_restart_upload:
                payload["forceRestartUpload"] = True
            if is_last and disable_stale_document_deletion_check:
                payload["disableStaleDocumentDeletionCheck"] = True

            log.info(
                "bulk_index_page",
                extra={
                    "upload_id": upload_id,
                    "page": i + 1,
                    "total_pages": len(pages),
                    "page_docs": len(page),
                    "is_first_page": is_first,
                    "is_last_page": is_last,
                },
            )
            results.append(self._post(url, payload, op="bulk_index_documents"))

        return results


class QueryClient(_BaseClient):
    """Reads via Search and Chat APIs using the Client (or Search) token."""

    def __init__(self, cfg: Config):
        super().__init__(cfg, token=cfg.token_for_search(), label="query")

    def search(
        self,
        query: str,
        page_size: int = 5,
        datasource: Optional[str] = None,
    ) -> CallResult:
        """Ranked retrieval for a natural-language query."""
        url = f"{self.cfg.rest_base_url}/search"
        payload: dict[str, Any] = {
            "query": query,
            "pageSize": page_size,
        }
        if datasource:
            payload["requestOptions"] = {
                "datasourcesFilter": [datasource],
            }
        return self._post(url, payload, op="search")

    def chat(self, message: str, context_docs: list[dict]) -> CallResult:
        """
        Grounded generation. We pass retrieved documents as inline context.

        Note: Chat can also retrieve against the tenant's indexed content
        on its own (ADR-001). We supply context explicitly here because the
        exercise requires using all three APIs.
        """
        url = f"{self.cfg.rest_base_url}/chat"
        payload = {
            "messages": [
                {
                    "author": "USER",
                    "messageType": "CONTENT",
                    "fragments": [{"text": message}],
                }
            ],
            "inlineDocs": context_docs,
        }
        return self._post(url, payload, op="chat")

File: server/glean-client.ts
/**
 * Thin HTTP clients for the Glean APIs, split by trust domain.
 *
 * IndexingClient — uses the Indexing token; writes to the knowledge graph.
 * QueryClient    — uses the Client token; reads via Search and Chat.
 *
 * This split mirrors Glean's own token model (separate Indexing vs Client tokens)
 * and makes the auth boundary obvious in code. See ADR-004 in DESIGN_NOTE.md.
 *
 * Responsibilities:
 * - Attach auth headers per client.
 * - Retry on transient failures (429, 5xx, network) with exponential backoff.
 * - Emit structured logs with per-call latency and request IDs.
 * - Surface non-transient errors clearly.
 */

import axios, { AxiosInstance, AxiosError } from 'axios';
import axiosRetry from 'axios-retry';
import { v4 as uuidv4 } from 'uuid';
import type { ConfiguredInstance } from './config.js';

export class GleanAPIError extends Error {
  constructor(
    public readonly status: number,
    public readonly body: string,
    public readonly requestId: string
  ) {
    super(`Glean API error ${status} (request_id=${requestId}): ${body.slice(0, 500)}`);
    this.name = 'GleanAPIError';
  }
}

export interface CallResult<T = any> {
  status: number;
  json: T;
  latencyMs: number;
  requestId: string;
}

class BaseClient {
  protected http: AxiosInstance;

  constructor(
    protected readonly cfg: ConfiguredInstance,
    token: string,
    protected readonly label: string
  ) {
    this.http = axios.create({
      timeout: cfg.config.httpTimeoutS * 1000,
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
        'User-Agent': 'glean-rag-node/1.0',
      },
    });

    // Configure retry logic for transient errors
    axiosRetry(this.http, {
      retries: cfg.config.retryMaxAttempts,
      retryDelay: axiosRetry.exponentialDelay,
      retryCondition: (error: AxiosError) => {
        // Retry on network errors or 429/5xx
        return (
          axiosRetry.isNetworkOrIdempotentRequestError(error) ||
          error.response?.status === 429 ||
          (error.response?.status !== undefined && error.response.status >= 500)
        );
      },
    });
  }

  protected async post<T = any>(
    url: string,
    payload: any,
    op: string
  ): Promise<CallResult<T>> {
    const requestId = uuidv4();
    const started = Date.now();

    try {
      const response = await this.http.post<T>(url, payload);
      const latencyMs = Date.now() - started;

      this.log('info', 'glean_api_call', {
        client: this.label,
        op,
        status: response.status,
        latencyMs,
        requestId,
      });

      return {
        status: response.status,
        json: response.data,
        latencyMs,
        requestId,
      };
    } catch (error) {
      const latencyMs = Date.now() - started;

      if (axios.isAxiosError(error)) {
        if (error.response) {
          // HTTP error response
          this.log('error', 'glean_api_error', {
            client: this.label,
            op,
            status: error.response.status,
            latencyMs,
            requestId,
            error: error.message,
          });

          throw new GleanAPIError(
            error.response.status,
            JSON.stringify(error.response.data),
            requestId
          );
        } else {
          // Network error
          this.log('warning', 'glean_api_network_error', {
            client: this.label,
            op,
            url,
            requestId,
            error: error.message,
          });
          throw error;
        }
      }

      throw error;
    }
  }

  protected log(level: string, message: string, extra: Record<string, any>): void {
    const logData = { message, ...extra };
    console.log(`[${level.toUpperCase()}]`, JSON.stringify(logData));
  }
}

export interface BulkIndexOptions {
  /** Page size for chunking documents across /bulkindexdocuments calls. */
  pageSize?: number;
  /** Override the auto-generated uploadId (must be stable across pages). */
  uploadId?: string;
  /** Restart any prior interrupted upload with this uploadId. */
  forceRestartUpload?: boolean;
  /**
   * Skip the safety check that prevents the last-page deletion sweep when
   * the bulk set is suspiciously smaller than the live index.
   */
  disableStaleDocumentDeletionCheck?: boolean;
  /** Override the configured datasource for this upload. */
  datasource?: string;
}

export class IndexingClient extends BaseClient {
  constructor(cfg: ConfiguredInstance) {
    super(cfg, cfg.config.gleanIndexingToken, 'indexing');
  }

  /**
   * Upsert documents to the configured datasource via /indexdocuments.
   * Additive — documents missing from this call are NOT removed. Use
   * bulkIndexDocuments() for full-sync semantics.
   * Note: Indexing is asynchronous — documents may not be immediately
   * searchable after this call returns 200.
   */
  async indexDocuments(
    documents: any[],
    opts: { datasource?: string } = {}
  ): Promise<CallResult> {
    const url = `${this.cfg.indexingBaseUrl}/indexdocuments`;
    const payload = {
      datasource: opts.datasource ?? this.cfg.config.gleanDatasource,
      documents,
      // uploadId makes re-runs idempotent on Glean's side too.
      uploadId: `prototype-${Math.floor(Date.now() / 1000)}`,
    };

    return this.post(url, payload, 'index_documents');
  }

  /**
   * Full-sync upload via /bulkindexdocuments. Chunks documents into pages
   * sharing one uploadId, marking isFirstPage / isLastPage. When the last
   * page lands, Glean drops any documents in the datasource that weren't
   * part of this upload — the canonical "replace the source" flow.
   *
   * Returns the per-page CallResults in upload order.
   */
  async bulkIndexDocuments(
    documents: any[],
    opts: BulkIndexOptions = {}
  ): Promise<CallResult[]> {
    const url = `${this.cfg.indexingBaseUrl}/bulkindexdocuments`;
    const pageSize = opts.pageSize ?? 50;
    const uploadId = opts.uploadId ?? `bulk-${Math.floor(Date.now() / 1000)}`;

    if (documents.length === 0) {
      throw new Error('bulkIndexDocuments: documents must be non-empty');
    }

    const pages: any[][] = [];
    for (let i = 0; i < documents.length; i += pageSize) {
      pages.push(documents.slice(i, i + pageSize));
    }

    const results: CallResult[] = [];
    for (let i = 0; i < pages.length; i++) {
      const isFirstPage = i === 0;
      const isLastPage = i === pages.length - 1;
      const payload: any = {
        datasource: opts.datasource ?? this.cfg.config.gleanDatasource,
        documents: pages[i],
        uploadId,
        isFirstPage,
        isLastPage,
      };
      if (isFirstPage && opts.forceRestartUpload) {
        payload.forceRestartUpload = true;
      }
      if (isLastPage && opts.disableStaleDocumentDeletionCheck) {
        payload.disableStaleDocumentDeletionCheck = true;
      }

      this.log('info', 'bulk_index_page', {
        uploadId,
        page: i + 1,
        totalPages: pages.length,
        pageDocs: pages[i].length,
        isFirstPage,
        isLastPage,
      });

      results.push(await this.post(url, payload, 'bulk_index_documents'));
    }

    return results;
  }
}

export class QueryClient extends BaseClient {
  constructor(cfg: ConfiguredInstance) {
    super(cfg, cfg.tokenForSearch(), 'query');
  }

  /**
   * Ranked retrieval for a natural-language query.
   */
  async search(
    query: string,
    pageSize: number = 5,
    datasource?: string
  ): Promise<CallResult> {
    const url = `${this.cfg.restBaseUrl}/search`;
    const payload: any = {
      query,
      pageSize,
    };

    if (datasource) {
      payload.requestOptions = {
        datasourcesFilter: [datasource],
      };
    }

    return this.post(url, payload, 'search');
  }

  /**
   * Get all indexed documents from a datasource.
   * Uses a broad search query to retrieve all available documents.
   */
  async getAllDocuments(datasource: string, maxResults: number = 100): Promise<CallResult> {
    const url = `${this.cfg.restBaseUrl}/search`;
    const payload: any = {
      query: '*', // Wildcard to match all documents
      pageSize: maxResults,
      requestOptions: {
        datasourcesFilter: [datasource],
      },
    };

    return this.post(url, payload, 'get_all_documents');
  }

  /**
   * Grounded generation. We pass retrieved documents as inline context.
   *
   * Note: Chat can also retrieve against the tenant's indexed content
   * on its own (ADR-001). We supply context explicitly here because the
   * exercise requires using all three APIs.
   */
  async chat(message: string, contextDocs: any[]): Promise<CallResult> {
    const url = `${this.cfg.restBaseUrl}/chat`;
    const payload = {
      messages: [
        {
          author: 'USER',
          messageType: 'CONTENT',
          fragments: [{ text: message }],
        },
      ],
      inlineDocs: contextDocs,
    };

    return this.post(url, payload, 'chat');
  }
}
